select_atoms accepts atomgroups

lilyminium opened the pull request:
Implements #2369 

Changes made in this Pull Request:
 - If `u.select_atoms` is given an Updating/AtomGroup, it returns that Updating/AtomGroup

@orbeckst commented [on the need for careful testing](https://github.com/MDAnalysis/mdanalysis/pull/2237#discussion_r272883354):

> I thought the same. Raise a separate issue. It looks like a simple change but this is such a fundamental method that it will require some careful testing. If it works consistently then it would be great. I think there was an old issue related to the "use AG where strings are allowed".

There is a simple test to check that it does return an AtomGroup, but otherwise I haven't got many ideas for others. Suggestions are welcome.

PR Checklist
------------
 - [ ] Tests? Kind of
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

richardjgowers commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -372,6 +372,14 @@ def test_global(self, universe):
+        ])
+    def test_accept_updatingatomgroup(self, universe, selstr):
+        ag = universe.select_atoms(selstr, updating=True)
+        assert ag is universe.select_atoms(ag)
```
So the ‘is’ statement here checks that they are literally the same object, but I think there’s (possible) implementations where it returns a copy. So the test might be better with ==

lilyminium replied:
I'm not sure I understand what you mean -- can checking `isinstance` modify the UpdatingAtomGroup?

https://github.com/MDAnalysis/mdanalysis/blob/cc71ab887bf5ee4609adef48a5ba8936ee65e963/package/MDAnalysis/core/groups.py#L2841-L2842

richardjgowers replied:
The test is using “assert A is B” not “assert A == B”, these are slightly different

orbeckst replied:
Perhaps have two tests: one that checks that you get the same object back (`is`) the other that it "is equal" (`==`). In this way we get noticed if our code does a copy somewhere where we don't expect it.

richardjgowers commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -363,6 +363,23 @@ def test_global(self, universe):
+        assert ag is universe.select_atoms(ag)
+    
+    @pytest.mark.parametrize('selstr',
+        ["around 4.0 bynum 1943",
```
The around statements are gonna be slow, especially if this is a real universe. Technically any selection (eg on names) with updating will be updating.

lilyminium replied:
Changed the selection string to 'resid 100'

richardjgowers commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -363,6 +363,23 @@ def test_global(self, universe):
         assert_equal(ag2.indices, ag1.indices)
+    
+    @pytest.mark.parametrize('selstr',
+        ["resname LYS and name NZ",
```
Parametrise seems unnecessary

lilyminium replied:
Removed extra tests.

orbeckst commented:
The macOS job ran into the 50min time limit. I restarted it but we might be back in the bad, old days where we have to run our tests faster...

orbeckst commented:
In terms of tests I was thinking about MDAnalysis.analysis.rms.RMSD, ... such tests would be added to the tests of the analysis code to make sure it works with selection strings _and_ AGs.

orbeckst commented:
Re: CHANGELOG: has to go into a 0.21.0

lilyminium commented:
Do we *want* `select_atoms` to return copies of AtomGroups?

What should happen when:

```
>>> ag = u.select_atoms('around 2 protein')
>>> u.select_atoms(ag, updating=True)
```
Should this generate an error or should the static AtomGroup be turned into a selection string of indices?

lilyminium commented:
`u.select_atoms(ag)` now returns the intersection of `u.atoms` and `ag`. I've added an `apply` method to AtomGroup that should probably be renamed to something private like `_apply_selection`, but for now it's the same as `Selection.apply`. 

I modified rms.RMSD to accept AtomGroups and added a couple tests. I'll probably need to go through all the analysis classes individually.

richardjgowers commented:
Yeah the updating case is weird because it won’t work as maybe it looks like at first. Once the ‘around ...’ selection is done, the resulting AG has no memory of how/why it was made, so can’t then retroactively be made updating.

I’m not super sold on the new way to perform intersections but I’ll think on it.

This whole patch is getting messy and I’m wondering if it’s better that analysis classes learn to accept either strings or (U)AGs.

orbeckst commented:
I agree that switching from selection strings to only (U)AGs is the cleaner way to do this. Given that this might be more complicated than anticipated then let's put this on hold for 2.0. There's no good reason to break lots of user code just because we find this to be cleaner/neater... at least not in the short term.

I see 1.0 as the "stable/old" MDAnalysis release (minus crud like flags #782 ...).

For 2.0 we can apply our lessons learned.

orbeckst commented:
I can be persuaded to change my opinion, of course. 

But I learned that a good way to save time is to say "no" and stop work while you can. @lilyminium sorry to vote for killing the PR at this stage after you put the work in but this work was necessary to show potential problems. And I think that other things have higher priority.

Again, I am happy to be disagreed with.

lilyminium commented:
Definitely happy to focus on other things. Not much work put in 😄 Analyses like rms.RMSD already accept both Universes and AtomGroups anyway, so it did feel a little unnecessary there. Should this be closed or left open?

orbeckst commented:
I’d close the PR (but don’t delete it) and leave the issue open. Add a quick note to the issue summarizing the discussion for future generations/future us, please. Thanks!

> Am 29.10.2019 um 00:28 schrieb Lily Wang <notifications@github.com>:
> 
> ﻿
> Definitely happy to focus on other things. Not much work put in 😄 Analyses like rms.RMSD already accept both Universes and AtomGroups anyway, so it did feel a little unnecessary there. Should this be closed or left open?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
